### DISCUSSION TITLE
how to redirect user back to the router came from after authentication

### DISCUSSION THREAD

**Discussion by ahakem** [vercel/next.js#13249] [Help] [answered]
Hello,
if the user opens a page which needs authentication I push him to the login page, but I need to push him back to this page
I achieved that in React using `react-router-dom`
but looking for a way in **NextJS**

I made that via a bad workaround but I think it's not a good idea.
`router.push('/signin', router.pathname)`
and on the other page

``` router.pathname !== router.asPath ? router.push(router.asPath) : router.push("/")```

this will open him the signing page and will keep the old URL,

any better answer please?

**Comment by vassbence** [comment]
You can handle this by having an url query called `next` for example and have the sign in page read it and redirect accordingly.

Let's say someone goes to `/dashboard` but is not logged in, the user gets redirected to `/login?next=%2Fdashboard` and in the login form submission you check for this `next` query and redirect the user back where he came from or to the default path if the query is not present.

**Reply by ahakem** [reply]
I'm looking for a way without passing a query in the URL, if possible

**Reply by ahakem** [reply]
I think I'll go with your answer as I could not find a better one. @vassbence

**Reply by EngrZaks** [reply]
One has to somehow, save the current route before redirecting. 
What I did when I couldn't find the right answer was to use localStorage. How is that? 😂

**Reply by tilda-deimos** [reply]
`router.back()` can also be used to go back in history https://nextjs.org/docs/api-reference/next/router#routerback

**Reply by EngrZaks** [reply]
No way router.back() can be the solution!
Shouldn't that be like going back to the login page? 🤔

**Reply by genevieveikechukwu** [reply]
@EngrZaks yours worked for me. thank you.

**Reply by lovishbansal00** [reply]
> No way router.back() can be the solution! Shouldn't that be like going back to the login page? 🤔

hey lets say this works, but i want to somehow maintain the state of my form as well.
suppose you are onboarding a new user and the onboarding has 2 pages (one form on each page), after directing from page 1 to page 2, my user wants to edit something on page 1 and goes back to page1. in this scenario how can i store the state of the form ?

**Reply by EngrZaks** [reply]
If your user is not done with onboarding, why route to the "next page" at all?
Use state management for the onboarding progress and then conditionally render different UI for each step of the unboarding.
This way, you can manage the whole thing better and once the unboarding is done, route to, say..... The dashboard home.

**Comment by msalahz** [comment]
Hi @ahakem , 
you can simply check in the begging of your pages if the user is logged in and if not, render login component instead of current page UI without changing the page URL and when the user login, make sure current page gets rendered, and in this case, the logged-in flag should be `false` and your page UI should render normally. You can follow this  [post](https://sergiodxa.com/articles/redirects-in-next-the-good-way/) for more details. I hope this helps.

**Reply by BernardA** [reply]
@msalahz That's a very poor solution.Why doesn't Next uses the same solution applied elsewhere, ie, allow a state property in the route object that, if present, would indicate that a redirect happened from a private route and which page to forward the user to. 
The login page also includes the layout ( header/footer), so you are saying we should render a page within a page - doubling header and footer or alternatively double all the login logic in a separate component.

**Reply by ahakem** [reply]
@BernardA, to be honest, it has totally different layout and different header and footer :D

**Comment by BernardA** [comment, marked as answer]
@ahakem Haven't tested yet, but this should work:

On the private route, do as indicated in [this issue](https://github.com/vercel/next.js/issues/4931): 

       componentDidMount() { // and/or componentDidUpdate, depending on what you're trying to do
          if (!auth) {
             Router.push({
                   pathname: '/login',
                   query: { from: {router.pathname} },
             }))
         }
       }

Then on the login page, after successful login: 

        handlePostLoginOk = () => {
              const { router } = this.props;
              if (router.query && router.query.from) {
                   router.push(router.query.from);
            }
      }

**Reply by ahakem** [reply]
will give it a try and let you know here @BernardA , tnx

**Reply by BernardA** [reply]
Please see edit

**Reply by ahakem** [reply]
@BernardA  this is the same way as mentioned by @vassbence and it's working fine, I just used `useEffect` as I'm using functional Component

**Reply by marvellousadesanya** [reply]
How does one implement this same code base in Next 13?

**Comment by BernardA** [comment]
Fine, I thought you were trying the not so good solution above. As I mentioned, in other frameworks you can add a property to the route and check it on the next component/page. Not sure why Next isn't using established proper solutions.

**Reply by ahakem** [reply]
you can do that in `react-router-dom` simply, but asking the same thing here why Next not providing that!.

**Reply by U-4-E-A** [reply]
I didn't see this discussion before and I have asked for the same in this [discussion](https://github.com/vercel/next.js/discussions/28184) (no replies yet). I am now looking at extending the NextRouter type definition to add this property myself.
